### PR TITLE
Simplify projection-of-point-on-polyline in contour.py.

### PR DESCRIPTION
**Pull request by anntzer**
## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/contour.py [hunk 1]
```diff
@@ -1333,7 +1305,7 @@ def find_nearest_contour(self, x, y, indices=None, pixel=True):
         xmin, ymin : float
             The point in the contour plot that is closest to ``(x, y)``.
         d : float
-            The distance from ``(xmin, ymin)`` to ``(x, y)``.
+            The squared distance from ``(xmin, ymin)`` to ``(x, y)``.
```
Comment (QuLogic) [inline lib/matplotlib/contour.py:1308]: This seems to be public, so I'm not sure we should change this.

File: lib/matplotlib/contour.py [hunk 2]
```diff
@@ -632,39 +607,36 @@ def _is_closed_polygon(X):
...
-            xcmin = xc
-            legmin = leg
-
-    return (dmin, xcmin, legmin)
+    if len(xys) == 1:
+        return (((p - xys[0]) ** 2).sum(), xys[0], (0, 0))
+    dxys = xys[1:] - xys[:-1]  # Individual segment vectors.
+    norms = (dxys ** 2).sum(axis=1)
+    norms[norms == 0] = 1  # For zero-length segment, replace 0/0 by 0/1.
+    rel_projs = np.clip(  # Proj. onto each segment in relative 0-1 coords.
```
Comment (QuLogic) [inline lib/matplotlib/contour.py:633]: The whole word fits.
```suggestion
    rel_projs = np.clip(  # Project onto each segment in relative 0-1 coords.
```

Comment (anntzer) [reply lib/matplotlib/contour.py]: yup

File: lib/matplotlib/contour.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/contour.py:1317]: I'm just describing the actual behavior, not changing it.

File: lib/matplotlib/contour.py [hunk 3]
```diff
@@ -633,39 +608,36 @@ def _is_closed_polygon(X):
...
+        Coordinates of point.
+
+    Returns
+    -------
+    d2min : float
+        Minimum square distance of *p* to *xys*.
+    proj : (float, float)
+        Projection of *p* onto *xys*.
+    imin : (int, int)
+        Consecutive indices of vertices of segment in *xys* where *proj* is.
```
Comment (timhoffm) [inline lib/matplotlib/contour.py:627]: Let's clarify the edge-cases
~~~
Segments are thought to include their end-points; i.e if the closest point
on the path is a node in *xys* with index *i*, this returns ``(i-1, i)``.
For the special case that *xys* is a single point, this returns ``(0, 0)``.
~~~

File: lib/matplotlib/contour.py [hunk 4]
```diff
@@ -633,39 +608,36 @@ def _is_closed_polygon(X):
...
+    if len(xys) == 1:
+        return (((p - xys[0]) ** 2).sum(), xys[0], (0, 0))
+    dxys = xys[1:] - xys[:-1]  # Individual segment vectors.
+    norms = (dxys ** 2).sum(axis=1)
+    norms[norms == 0] = 1  # For zero-length segment, replace 0/0 by 0/1.
+    rel_projs = np.clip(  # Project onto each segment in relative 0-1 coords.
+        ((p - xys[:-1]) * dxys).sum(axis=1) / norms,
+        0, 1)[:, None]
+    projs = xys[:-1] + rel_projs * dxys  # Projs. onto each segment, in (x, y).
+    d2s = ((projs - p) ** 2).sum(axis=1)  # Distances.
```
Comment (timhoffm) [inline lib/matplotlib/contour.py:638]:
```suggestion
    d2s = ((projs - p) ** 2).sum(axis=1)  # Squared distances.
```

File: lib/matplotlib/contour.py (hunk 3 above)
Comment (anntzer) [reply lib/matplotlib/contour.py:627]: yup

File: lib/matplotlib/contour.py (hunk 4 above)
Comment (anntzer) [reply lib/matplotlib/contour.py]: yup

**Pull request merged**